### PR TITLE
fix authz(rbac rules) builder

### PR DESCRIPTION
**Pull request by zengyuxing007**
**Please provide a description of this PR:**

The authz policy configuration that was expected to work only on http(hcm) was incorrectly applied to the tcp filter, causing tcp requests to be incorrectly blocked.

Fixes [ #44348 ]

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [X] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/security/authz/model/model.go [hunk 1]
```diff
@@ -201,6 +201,10 @@ func generatePermission(rl ruleList, forTCP bool, action rbacpb.RBAC_Action) (*r
 		}
 		and = append(and, ret...)
 	}
+	if action == rbacpb.RBAC_DENY {
```
Comment (hanxiaop) [inline pilot/pkg/security/authz/model/model.go:204]: It looks like this will break the current behavior

Comment (hanxiaop) [reply pilot/pkg/security/authz/model/model.go:204]: See this warning https://github.com/istio/istio/blob/master/pkg/config/validation/validation.go#L2120-L2125

Comment (zengyuxing007) [reply pilot/pkg/security/authz/model/model.go]: > See this warning https://github.com/istio/istio/blob/master/pkg/config/validation/validation.go#L2120-L2125

Declaring that it only works for a specific port is one way to do this, but obviously this increases the cost of understanding for the user and a better way would be to:

Rules in effect for http only should not be applied to tcp filters (envoy.filters.network.rbac)

**Review by howardjohn — CHANGES_REQUESTED** [review]
Thanks for the PR! Given this is a super highly sensitive area we are going to need tests covering this before proceeding

**Comment by zengyuxing007** [comment]
> #44348

Thank you for your reply, I will add or fix the relevant tests.

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 2]
```diff
@@ -355,6 +355,7 @@ func verify(t *testing.T, gots []proto.Message, baseDir string, wants []string,
 
 		util.RefreshGoldenFile(t, []byte(gotYaml), wantFile)
 		if err := util.Compare([]byte(gotYaml), []byte(wantYaml)); err != nil {
+			t.Errorf("goYaml:%s\n", gotYaml)
```
Comment (ymesika) [inline pilot/pkg/security/authz/builder/builder_test.go:358]: Is this a debug message leftover?

Comment (zengyuxing007) [reply pilot/pkg/security/authz/builder/builder_test.go:358]: No, only if the test case fails

**Comment by howardjohn** [comment]
Thanks for the PR. As noted in https://github.com/istio/istio/issues/44348#issuecomment-1506082112, this behavior is working as intended. In the past few months we had many hours of debate about this functionality and decided to keep it as-is, so I don't think this change will be accepted. We can continue on the issue discussion of other changes, such as improved docs, etc. Thanks!

**Pull request closed without merging**